test/probes/morgan.test.js

```diff
@@ -323,6 +349,127 @@ describe(`probes.morgan ${version}`, function () {
     })
   })
 
+  //
+  // for each mode verify that insert works with format string
+  //
+  insertModes.forEach(mode => {
+    const maybe = mode === false ? 'not ' : ''
+    eventInfo = undefined
+
+    it(`should ${maybe}insert in sync when mode=${mode} using a format string`, function (done) {
+      let traceId
+
+      // this gets reset in beforeEach() so set it in the test.
+      ao.cfg.insertTraceIdsIntoLogs = mode
+      logger = makeLogger(':method :trace :url :status :res[content-length]')
+
+      function localDone () {
+        checkEventInfo(eventInfo, fakeReq, fakeRes, mode === false ? undefined : traceId)
+        done()
+      }
+
+      helper.test(emitter, function (done) {
+        ao.instrument(spanName, function () {
+          traceId = ao.requestStore.get('topSpan').events.exit.event.toString()
+          // log
+          logger(fakeReq, fakeRes, function (err) {
+            expect(err).not.ok
+          })
+          fakeRes.writeHead(200)
+          fakeRes.finished = true
+        })
+        done()
+      }, [
+        function (msg) {
+          msg.should.have.property('Layer', spanName)
+          msg.should.have.property('Label', 'entry')
+        },
+        function (msg) {
+          msg.should.have.property('Layer', spanName)
+          msg.should.have.property('Label', 'exit')
+        }
+      ], localDone)
+    })
+  })
+
+  //
+  // for each mode verify that insert works with format string when no insertion is requested
+  //
+  insertModes.forEach(mode => {
+    eventInfo = undefined
+
+    it(`should never insert in sync when mode=${mode} using a format string and no insertion is requested`, function (done) {
+      // this gets reset in beforeEach() so set it in the test.
+      ao.cfg.insertTraceIdsIntoLogs = mode
+      logger = makeLogger(':method :url :status :res[content-length]')
+
+      function localDone () {
+        checkEventInfo(eventInfo, fakeReq, fakeRes, null)
+        done()
+      }
+
+      helper.test(emitter, function (done) {
+        ao.instrument(spanName, function () {
+          // log
+          logger(fakeReq, fakeRes, function (err) {
+            expect(err).not.ok
+          })
+          fakeRes.writeHead(200)
+          fakeRes.finished = true
+        })
+        done()
+      }, [
+        function (msg) {
+          msg.should.have.property('Layer', spanName)
+          msg.should.have.property('Label', 'entry')
+        },
+        function (msg) {
+          msg.should.have.property('Layer', spanName)
+          msg.should.have.property('Label', 'exit')
+        }
+      ], localDone)
+    })
+  })
+
+  //
+  // for each mode verify that insert works with format function
```

Comment:
To confirm, the test case below verifies if a custom function is used to define the log message but our trace context is not requested (via using the `tokens` property) we don't insert.

Can we add a test case where the custom function uses the `tokens.trace()` to verify correct insertion if the trace context is requested?

Reply:
Indeed. Will do.